pkg/test_framework/helpers.go

```diff
@@ -262,8 +263,150 @@ func getReady(obj runtime.Object) (int32, error) {
 			log.Panic(err)
 		}
 		return int32(len(list.Items)), nil
+	case int64:
+		return int32(t), nil
+	case int32:
+		return t, nil
+	case int:
+		return int32(t),nil
 	default:
-		log.Panicf("Unsupported type %v", t.GetObjectKind())
+		log.Panicf("Unsupported type %t", obj)
 	}
 	return 0, nil
 }
+
+type WaitForResourceFn func(obj unstructured.Unstructured)(int, error)
+
+// Wait for a singular resource like the one provided. Will use APIVersion and Kind from
+// like parameter. If name parameter is provided, will use Name and Namespace from name
+// parameter, otherwise will use Name and Namespace from like as well.
+// Optionally takes a function which can be used to transform a single resource into
+// a number (e.g. acting on some part of the object's spec or status section).
+func WaitForResource(wanted int, timeout time.Duration, client1 client.Client, like runtime.Object, name runtime.Object, fn WaitForResourceFn) error {
+	if name == nil {
+		name = like
+	}
+	apiver, kind := like.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
+	objname, objns := getSingularNameNs(name)
+	reps := int32(wanted)
+
+	return WaitFor(
+		&reps,
+		timeout,
+		func() (interface{}, error) {
+			res := &unstructured.Unstructured{}
+			res.SetAPIVersion(apiver)
+			res.SetKind(kind)
+			err := client1.Get(context.TODO(), client.ObjectKey{
+				Namespace: objns,
+				Name: objname,
+			}, res)
+			if err != nil {
+				return nil, err
+			}
+			if fn == nil {
+				return res, nil
+			}
+			return fn(*res)
+		},
+	)
+}
+
+type WaitForResourcesFn func(obj unstructured.UnstructuredList)(int, error)
+
+// Wait for a list of resources like the one provided. Will use APIVersion and Kind from
+// like parameter. If name parameter is provided, will use Namespace (only!) from name
+// parameter, otherwise will use Namespace from like as well.
+// Optionally takes a function which can be used to transform the list into
+// a number (e.g. filtering by some part of the objects' name or status).
+//
+// Note: lists of kubernetes resources do not carry a singular Namespace, so if name
+// is nil AND like is a list of objects, you will get a panic from getSingularNameNs.
```

Comment:
Any reason for not giving the responsibility of panic to `WaitForResources` as opposed to giving it to `getSingularNameNs`.  By the definition and contract of func `getSingularNameNs` it should not be passed list of objects. We can just panic from this function directly and forcing the contract of this function to take only singular object as opposed to list.
Any reason for not handling this case in following way : if `like` is a list as opposed to singular object, just use 0 indexed element of that list.

Reply:
The reason for 1. is code cohesion; if `WaitForResources` was to make that decision then it would have to own a list of types from which we are able (or not) to read a Namespace, which means duplication of logic implemented inside `getSingularNameNs`. Such duplication is bad design, so let's not do that; instead panic will show the stacktrace with all the information the user needs to fix their test.
The reason for 2. is to avoid a condition where failing (panicking) test is subject to some runtime condition (i.e. do we have any items at all in the list), and hence more difficult to reproduce. Also, in case the user was to write a most natural (but incorrect) code passing an empty list, the panic would indicate that there is an empty collection of items, rather than an invalid type was used. Which would push them to more complex and clumsy workaround of creating a list object and adding a dummy element, rather than simply use a non-list object.